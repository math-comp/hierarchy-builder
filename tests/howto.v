From Coq Require Import ZArith ssrfun ssreflect.
From HB Require Import structures.
From HB Require Import demo1.hierarchy_5.

HB.howto Ring.type.

HB.howto Ring.type 2.

HB.howto Z Ring.type.

HB.howto Z Ring.type 2.

Fail HB.howto Z Ring.type 0.

HB.howto AddComoid.type Ring.type.

HB.instance
  Definition _ :=
  AddAG_of_TYPE.Build Z 0%Z Z.add Z.opp
    Z.add_assoc Z.add_comm Z.add_0_l Z.add_opp_diag_l.

HB.howto Z Ring.type.

HB.howto AddAG.type Ring.type.

HB.instance
  Definition _ :=
  Ring_of_TYPE.Build Z 0%Z 1%Z Z.add Z.opp Z.mul
    Z.add_assoc Z.add_comm Z.add_0_l Z.add_opp_diag_l
    Z.mul_assoc Z.mul_1_l Z.mul_1_r
    Z.mul_add_distr_r Z.mul_add_distr_l.

HB.howto Z Ring.type.

Fail HB.howto AddMonoid_of_TYPE.
Fail HB.howto Ring_of_TYPE.
Fail HB.howto Ring.
