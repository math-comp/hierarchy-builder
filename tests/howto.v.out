HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - Ring_of_TYPE
    - AddAG_of_TYPE; BiNearRing_of_AddMonoid
    - AddAG_of_TYPE; Ring_of_AddAG
    - AddAG_of_TYPE; SemiRing_of_AddComoid
    - AddComoid_of_TYPE; Ring_of_AddComoid
    - AddComoid_of_TYPE; AddAG_of_AddComoid; BiNearRing_of_AddMonoid
    - AddComoid_of_TYPE; AddAG_of_AddComoid; Ring_of_AddAG
    - AddComoid_of_TYPE; AddAG_of_AddComoid; SemiRing_of_AddComoid
    - AddMonoid_of_TYPE; AddComoid_of_AddMonoid; Ring_of_AddComoid

HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - Ring_of_TYPE
    - AddAG_of_TYPE; BiNearRing_of_AddMonoid
    - AddAG_of_TYPE; Ring_of_AddAG
    - AddAG_of_TYPE; SemiRing_of_AddComoid
    - AddComoid_of_TYPE; Ring_of_AddComoid

HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - Ring_of_TYPE
    - AddAG_of_TYPE; BiNearRing_of_AddMonoid
    - AddAG_of_TYPE; Ring_of_AddAG
    - AddAG_of_TYPE; SemiRing_of_AddComoid
    - AddComoid_of_TYPE; Ring_of_AddComoid
    - AddComoid_of_TYPE; AddAG_of_AddComoid; BiNearRing_of_AddMonoid
    - AddComoid_of_TYPE; AddAG_of_AddComoid; Ring_of_AddAG
    - AddComoid_of_TYPE; AddAG_of_AddComoid; SemiRing_of_AddComoid
    - AddMonoid_of_TYPE; AddComoid_of_AddMonoid; Ring_of_AddComoid

HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - Ring_of_TYPE
    - AddAG_of_TYPE; BiNearRing_of_AddMonoid
    - AddAG_of_TYPE; Ring_of_AddAG
    - AddAG_of_TYPE; SemiRing_of_AddComoid
    - AddComoid_of_TYPE; Ring_of_AddComoid

The command has indeed failed with message:
HB: no solution found, try to increase search depth.
HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - Ring_of_AddComoid
    - AddAG_of_AddComoid; BiNearRing_of_AddMonoid
    - AddAG_of_AddComoid; Ring_of_AddAG
    - AddAG_of_AddComoid; SemiRing_of_AddComoid

HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - BiNearRing_of_AddMonoid
    - Ring_of_AddAG
    - SemiRing_of_AddComoid

HB: solutions (use 'HB.about F.Build' to see the arguments of each factory F):
    - BiNearRing_of_AddMonoid
    - Ring_of_AddAG
    - SemiRing_of_AddComoid

HB: nothing to do.
